app/src/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -614,34 +526,233 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
+        // We want to save our final image file in the instance folder for this form, so that it
+        // gets sent to HQ with the form
+        String instanceFolder =
+                mInstancePath.substring(0, mInstancePath.lastIndexOf("/") + 1);
+        String imageFilename = System.currentTimeMillis() + ".jpg";
+        String finalFilePath = instanceFolder + imageFilename;
+
+        boolean savedScaledImage = false;
+        if (shouldScale) {
+            savedScaledImage = scaleImage(originalImage, finalFilePath);
+        }
+
+        File fileToReturn;
+        if (!savedScaledImage) {
+            // If we didn't create a scaled image and save it to the final path, then relocate the
+            // original image from the temp filepath to our final path
+            File finalFile = new File(finalFilePath);
+            if (!originalImage.renameTo(finalFile)) {
+                fileToReturn = originalImage;
+                Log.e(TAG, "Failed to rename " + originalImage.getAbsolutePath());
+            } else {
+                fileToReturn = finalFile;
+                Log.i(TAG, "renamed " + originalImage.getAbsolutePath() + " to " + finalFile.getAbsolutePath());
+            }
+        } else {
+            // Otherwise, relocate the original image to a raw/ folder, so that we still have access
+            // to the unmodified version
+            String rawDirPath = instanceFolder + "/raw";
+            File rawDir = new File(rawDirPath);
+            if (!rawDir.exists()) {
+                rawDir.mkdir();
+            }
+            File rawImageFile = new File (rawDirPath + "/" + imageFilename);
+            if (!originalImage.renameTo(rawImageFile)) {
+                fileToReturn = originalImage;
+                Log.e(TAG, "Failed to rename " + originalImage.getAbsolutePath());
+            } else {
+                fileToReturn = rawImageFile;
+                Log.i(TAG, "renamed " + originalImage.getAbsolutePath() + " to " + rawImageFile.getAbsolutePath());
+            }
+        }
+        return fileToReturn;
+    }
 
-        //Ugh, copied from the odkview mostly, that's stupid
-        for(QuestionWidget q : ((ODKView)mCurrentView).getWidgets()) {
-            //Figure out if we have a pending intent widget
-            if (q instanceof IntentWidget) {
-                if(((IntentWidget) q).getFormId().equals(mFormController.getPendingCalloutFormIndex())) {
-                    bestMatch = (IntentWidget)q;
+    /**
+     * Attempts to scale down an image file based on the max dimension in the current ImageWidget.
+     * If a max dimension is set, and at least one of the dimensions of the original image exceeds
+     * that maximum, then make the larger side's dimension equal to the max dimension, and scale
+     * down the smaller side such that the original aspect ratio is maintained
+     */
+    private boolean scaleImage(File originalImage, String finalFilePath) {
```

Comment:
This file is huge. Is it possible to extract the image scaling methods you're adding here into another file?